### ISSUE TITLE
Add a note about system setup

### ISSUE THREAD

**Issue by rubak** [eddelbuettel/prrd#13]
As I mentioned in another thread it would be great with authoritative advice on how to setup the machine **before** running `prrd`. Do you setup a revdep docker image for each package you maintain with the needed system libraries? Or do you have all the needed libraries across packages installed on a single machine and then run the checks on that? It could be valuable to have a note about this somewhere like the README of `prrd` to let people know that there is an important step here and there might not be an easy solution for the setup step.

Did you ever look at Jeroen's [maketools vignette](https://cran.r-project.org/web/packages/maketools/vignettes/sysdeps.html) about runtime/compiletime dependencies etc? Would it be possible to combine docker, `maketools`, bspm, apt and friends to make a recipe for how to approach this problem in general?

**Comment by eddelbuettel** [comment]
You beat me to opening an issue here. :)

There are a few issues, intertwined:
- it is a complex issue, and a somewhat complex package
- total "worldwise" user base is unknown, likely small, maybe less fingers on one hand ?
- once setup it just runs ...
- combine those points and you have near zero recurrent setups
- so nobody works on setup documentation :-/

What I (currently) do (on the machine / VM where I am guest with my `prrd` use) is to have a
- 'directory for test library' in `~/tmp/lib` where all needed-for-prrd packages are installed
- 'directory for tests' in `~/tmp/prrd`
- all system dependencies are just that: system
- a black list of maybe 40 or so packages (compared to 2200+ package) which I cannot or do not want to [1] test

So I guess we cannot all mutually wait on each other for documentation to "magically" appear.  Someone has to write up some notes.  Maybe you can do that for your setup?  Maybe I can for mine?  Maybe we both do it?

[1] Some are antisocial and run for 30 minutes, depend on too much, or are otherwise annoying.  The blacklist is public.

**Comment by eddelbuettel** [comment]
> Would it be possible to combine docker, maketools, bspm, apt and friends to make a recipe for how to approach this problem in general?

Sure, if you want to. To me, `bspm` is entirely orthogonal.  Because I need "everything", more or less, I actually prefer source in this use case.

**Comment by eddelbuettel** [comment]
I had not seen his [maketools vignette](https://cran.r-project.org/web/packages/maketools/vignettes/sysdeps.html) but I wrote a similar tool myself when RSPM was new as one needs to guestimate the system depends in which they don't (really) supply.  I think I left mine at a working protype and moved on to BSPM.

[ Goes looking for three minutes ]

I had actually forgotten what I called it but the package and repo are here:  https://github.com/eddelbuettel/chshli  Looks like it only had six commits over two months in two distinct bursts...

**Comment by rubak** [comment]
> The blacklist is public.

Ouch. I like that...

I'm only co-author and not maintainer on `spatstat` and have not done revdep checks until recently, where we started working on a CRAN mandated split-up of the package which would introduce breaking changes for ~100 revdeps. However recursive dependenices ring in at thousands of packages, so I want to check 100 source packages which requires:

1. whatever system headers/libraries they use 
2. thousands of other R packages

As a complete newbie I pulled rocker/r-bspm:20.04 and tried to install as many packages as possible via `bspm` to save time and let the package manager figure out as many system dependencies as possible for packages that were only run/check-time dependencies of my revdeps. Afterwards, installing the remaining packages from source gave occasional hiccups due to missing headers, and I then Googled and installed appropriate headers with apt (without really taking notes as I was just frantically trying to get things working and wondering if this was a good workflow, so I cannot reproduce what I did). In the end I had a 11 GB docker image which was able to run reverse dependency checks and discover that almost all packages failed with the new version as expected.

Your point "once setup it just runs..." was actually what I was worried about when I chose to try to run this in docker. What if I want to move my revdep checks to run on a department server instead of my laptop? What happens when I wipe my machine or get a new one? Do I then have to start all over? Things like an R library and system packages is not what I usually have in my backups. However, I'm new to docker so I actually don't even know the good way to share this from one machine to the other. Should I really have noted a bunch of commands in a dockerfile or should I just copy the image? Is it even meaningful to put this in docker?

I don't expect I have found a great way to do this and it was because it all felt so ad hoc I was hoping that there existed some advice from experienced package developers that did this before me in a larger scale. Maybe the "target audience" is simply too small for this to make sense, and people just have to cook up a method that works for them as they go along.

**Comment by eddelbuettel** [comment]
(I am so sorry. I missed your reply for three days, but was pondering this in the meantime.)

(And I just had to install `s2` on my rev.dep box so yes, I understand :) )

First off, CRAN's new request for recursive rev.deps I ... mostly have to ignore as Rcpp's _direct_ reverse depends are around 2270 today and take a day.  That said, for smaller packages, one probably could. I think I did a test once. In any event, that is just _one call_ to `tools::package_dependencies()` ...

Next, Docker vs full machine. You would be surprised to hear that I ... actually only ever use the latter.  I think years ago I started doing it on my (beefy at the time, many cores, lots of ram) workstation at work.  As mentioned above, I just parcel off a directory for additional packages I only install for rev.dep checks, and `prrd` supports that.  _As I generally am on Debian testing or Ubuntu 'current'_ I have always maintained that I can do as CRAN as trust all system libraries_.  I still believe this holds (_modulo_ maybe the fact that I do not primarily work with the `proj`/`gdal` complex -- dunno) and I have _not_ been bitten.  I also tend to install quite a bit on my main development machines.   Docker images, as you saw, will easily 'swell' and it is an open computational problem how to partition them into working disjoint sets.  For another time or project...  So in sum, maybe a dedicated machine is ideal (and I have access to one now, though it is old-ish and slow), a large departmental machine is still possible.  There is also our old Unix friend `chroot` (which we use a ton inside Debian for different build ens -- there was a world before Docker too...)

I think your call for action on documentation is a good one.  How about we try this jointly via new vignette?  If we move this from "essentially NOT documented" to "under documented" it can rest there for a bit and in a second phase we may have energy to move from "under documented" to "somewhat documented" :)

**Comment by rubak** [comment]
I agree that "under documented" is a good place to aim for right now :) A new vignette may be a good way to do this. I guess the workflow is that I fork and make a new branch with an empty vignette and then make a PR with that, so we both can make commits on that PR? Or is there a more clever way?

Maybe we are talking about different things when thinking about recursive dependencies. Are you saying that CRAN asks us to find all recursive dependencies with `tools::package_dependencies(..., reverse = TRUE, recursive = TRUE)` and then test all these packages with the new version of our own package? That sounds scary... My point about needing many other packages was that simply to run a full `R CMD check` on my direct rev deps I need to install all the direct dependencies (incl. `Suggests`) of those packages for the check to be able to run (and then recursively the strong dependencies from there). Thus in my understanding what I need is found like this:

```r
revdep <- tools::package_dependencies("spatstat", which = "all", reverse = TRUE)[[1]]
revdep1 <- tools::package_dependencies(revdep, which = "all")
revdep1 <- Reduce(union, revdep1)
revdep2 <- tools::package_dependencies(revdep1, which = c("Depends", "Imports", "LinkingTo"), recursive=TRUE)
revdep2 <- Reduce(union, revdep2)
revdepX <- union(revdep1, revdep2)
total <- union(revdep, revdepX)
```
In an ideal world with binaries available for everything I would:

1. install binaries of `total`
2. install system libraries needed for `revdep`
3. run `R CMD check` on `revdep`

I had done things slightly wrong before where I had to many levels of suggests and/or recursiveness and ended up with 2000+ needed packages for a spatstat revdep check. If my thinking here is correct I currently get 800+ which may decrease soon as the package is split up and revdeps move to various sub-packages.

**Comment by eddelbuettel** [comment]
Re workflow:  Many ways to skin the cat.  We could just create a new mini-repo where we both can write, and then "eventually" move the vignette here, similar to repos for papers.

Re scary: Indeed. But that is what R Core recently said. Discussion (e.g. with Mark vdL) was on some list somewhere.  Less scary: I trim where I can and omit a) all Suggests where I can (some packages force me to load them as they do not test) and skip b) vignettes.

Re depends:  There is code in `prrd` to compile this (`enqueueDepends`), but I don't currently use it,  You could poke there.

Re 2000+:  That can happen. After all we have 17k packages at CRAN.
